File: src/main/java/aurocosh/divinefavor/common/item/talismans/spell/base/SpellTalismanSummonEntity.java
package aurocosh.divinefavor.common.item.talismans.spell.base;

import aurocosh.divinefavor.common.favor.ModFavor;
import net.minecraft.entity.EntityLiving;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.DifficultyInstance;
import net.minecraft.world.World;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.EnumSet;

public class SpellTalismanSummonEntity<T extends EntityLiving> extends ItemSpellTalisman {
    private final Class<? extends T> clazz;

    public SpellTalismanSummonEntity(String name, ModFavor favor, int favorCost, EnumSet<SpellOptions> options, Class<? extends T> clazz) {
        super(name, favor, favorCost, options);
        this.clazz = clazz;
    }

    @Override
    protected void performActionServer(TalismanContext context) {
        try {
            World world = context.world;
            BlockPos spawnPos = getPosition(context);

            Constructor<? extends T> constructor = clazz.getConstructor(World.class);
            T entityLiving = constructor.newInstance(world);
            entityLiving.setLocationAndAngles(spawnPos.getX(), spawnPos.getY(), spawnPos.getZ(), 0, 0.0F);

            DifficultyInstance difficulty = entityLiving.world.getDifficultyForLocation(entityLiving.getPosition());
            entityLiving.onInitialSpawn(difficulty, null);

            preProcessEntity(entityLiving, context);
            world.spawnEntity(entityLiving);
            postProcessEntity(entityLiving, context);
        }
        catch (InvocationTargetException | NoSuchMethodException | InstantiationException | IllegalAccessException e) {
            e.printStackTrace();
        }
    }

    protected BlockPos getPosition(TalismanContext context) {
        return context.pos.offset(context.facing);
    }

    protected void preProcessEntity(T entityLiving, TalismanContext context) {
    }

    protected void postProcessEntity(T entityLiving, TalismanContext context) {
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/item/talismans/spell/base/SpellTalismanSummonEntity.kt
package aurocosh.divinefavor.common.item.talismans.spell.base

import aurocosh.divinefavor.common.lib.extensions.multicatch
import aurocosh.divinefavor.common.spirit.base.ModSpirit
import net.minecraft.entity.EntityLiving
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import java.lang.reflect.InvocationTargetException
import java.util.*

open class SpellTalismanSummonEntity<T : EntityLiving>(name: String, spirit: ModSpirit, favorCost: Int, options: EnumSet<SpellOptions>, private val clazz: Class<out T>) : ItemSpellTalisman(name, spirit, favorCost, options) {

    override fun performActionServer(context: TalismanContext) {
        try {
            val world = context.world
            val spawnPos = getPosition(context)

            val constructor = clazz.getConstructor(World::class.java)
            val entityLiving = constructor.newInstance(world)
            entityLiving.setLocationAndAngles(spawnPos.x.toDouble(), spawnPos.y.toDouble(), spawnPos.z.toDouble(), 0f, 0.0f)

            val difficulty = entityLiving.world.getDifficultyForLocation(entityLiving.position)
            entityLiving.onInitialSpawn(difficulty, null)

            preProcessEntity(entityLiving, context)
            world.spawnEntity(entityLiving)
            postProcessEntity(entityLiving, context)
        } catch (e: Throwable) {
            e.multicatch(InvocationTargetException::class, NoSuchMethodException::class, InstantiationException::class, IllegalAccessException::class) {
                e.printStackTrace()
            }
        }
    }

    protected fun getPosition(context: TalismanContext): BlockPos {
        return context.pos.offset(context.facing)
    }

    protected fun preProcessEntity(entityLiving: T, context: TalismanContext) {}

    protected open fun postProcessEntity(entityLiving: T, context: TalismanContext) {}
}
